fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/tree/lowcardinality/DecodeCollector.java

```diff
@@ -0,0 +1,611 @@
+            return visitPhysicalJoin(context.parent, context);
+        }
+        return result;
+    }
+
+    @Override
+    public DecodeNodeInfo visitPhysicalOlapScan(OptExpression optExpression, DecodeNodeInfo context) {
+        PhysicalOlapScanOperator scan = optExpression.getOp().cast();
+        OlapTable table = (OlapTable) scan.getTable();
+        long version =
+                table.getPartitions().stream().map(Partition::getVisibleVersionTime).max(Long::compareTo).orElse(0L);
+
+        if ((table.getKeysType().equals(KeysType.PRIMARY_KEYS))) {
+            return DecodeNodeInfo.EMPTY;
+        }
+        if (table.hasForbitGlobalDict()) {
+            return DecodeNodeInfo.EMPTY;
+        }
+
+        // check dict column
+        DecodeNodeInfo info = new DecodeNodeInfo();
+        for (ColumnRefOperator column : scan.getColRefToColumnMetaMap().keySet()) {
+            if (!column.getType().isVarchar()) {
+                continue;
+            }
+
+            ColumnStatistic columnStatistic =
+                    GlobalStateMgr.getCurrentStatisticStorage().getColumnStatistic(table, column.getName());
+            // Condition 2: the varchar column is low cardinality string column
+            if (!FeConstants.USE_MOCK_DICT_MANAGER && (columnStatistic.isUnknown() ||
+                    columnStatistic.getDistinctValuesCount() > CacheDictManager.LOW_CARDINALITY_THRESHOLD)) {
+                LOG.debug("{} isn't low cardinality string column", column.getName());
+                continue;
+            }
+
+            // Condition 3: the varchar column has collected global dict
+            if (!IDictManager.getInstance().hasGlobalDict(table.getId(), column.getName(), version)) {
+                LOG.debug("{} doesn't have global dict", column.getName());
+                continue;
+            }
+
+            Optional<ColumnDict> dict = IDictManager.getInstance().getGlobalDict(table.getId(), column.getName());
+            // cache reaches capacity limit, randomly eliminate some keys
+            // then we will get an empty dictionary.
+            if (dict.isEmpty()) {
+                continue;
+            }
+
+            info.outputStringColumns.union(column);
+            info.inputStringColumns.union(column);
+            stringRefToDefineExprMap.put(column.getId(), column);
+            scanStringColumns.add(column.getId());
+            expressionStringRefCounter.put(column.getId(), 0);
+            globalDicts.put(column.getId(), dict.get());
+        }
+
+        if (info.outputStringColumns.isEmpty()) {
+            return DecodeNodeInfo.EMPTY;
+        }
+
+        return info;
+    }
+
+    private void collectPredicate(Operator operator, DecodeNodeInfo info) {
+        if (operator.getPredicate() == null) {
+            return;
+        }
+        DictExpressionCollector dictExpressionCollector = new DictExpressionCollector(info.outputStringColumns);
```

Comment:
Predicate is compoundAnd. 
what's the rewrite result for
```
A > 'XXX' and A = B and if (A = 'x', 0, 1)
```

Reply:
Dict(A > 'XXX' and A = B and if (A = 'x', 0, 1)) and Dict(A) = rand()